### DISCUSSION TITLE
Handling both Serial & USB

### DISCUSSION THREAD

**Discussion by tercerapersona** [lathoub/Arduino-USBMIDI#21] [Q&A]
Hi, I'm working in a MIDI instrument that I would like to support both serial & midi clock + messages, like any other industrial instrument.

I haven't seen any example on the whole internet, would you provide one?

Use cases:
- MIDI clock read from serial is connected.
- MIDI clock read from USB is connected.
- MIDI clock read from digital pin (optional, audio cable 3.5mm)
- MIDI messages read from serial & USB (if any or both are connected)
- MIDI messages sent to serial & USB (if any or both are connected)

**Comment by lathoub** [comment]
[moved from Issues]

> MIDI clock read from serial is connected.

What do you mean with 'is connected'? ('*if*' connected?) The MIDI Serial protocol does not keep a connection status.

> MIDI clock read from digital pin (optional, audio cable 3.5mm)

I believe this is the same as serial MIDI (audio cable 3.5mm jack is just a small serial MIDI connector)

From what I understand you can create 1 instance for Serial MIDI and another for USB midi and attached call-backs to both (see the examples). Then send messages from your instrument to both Serial and USB.

**Comment by tercerapersona** [comment]
Thanks! yeah, sending MIDI might be trivial, just call both send methods.

However MIDI clock is a bit tricky. I guess I should register both callbacks (MIDI & Serial) and have some bool flag to switch between both. 

Same with the 3.5mm audio jack cable, which I meant that 5v pulse clock (24 beat per quarter) coming from Korg Volca series or Eurorack modules.

It's a bit challenging as all examples on the internet either use USB or serial. :sweat_smile:

**Comment by serlecu** [comment]
I don't really understand if the origin of this discussion was more about incoming MIDI, but actually, I'm having problems with the implementation of midi USB+DIN output, more specifically with CC messages. 

All examples that send any midi work for me, but none of them implement USB and DIN output simultaneously. Also, with my own code it works only with one or the other output method flawlessly but never both outputs together.

I create the "DIN" instance like this:
```
MIDI_CREATE_INSTANCE(HardwareSerial, Serial1, MIDIhw); 
```
And the USB instance like this (you can see I gave it cable#1, bcs at the beginning I was afraid #0 would interfere):
```
USBMIDI_CREATE_INSTANCE(1, MIDIusb)
```

Mainly following this guess of yours, @lathoub, about creating separate instances and different sends for each (the call-backs I don't need them):
> From what I understand you can create 1 instance for Serial MIDI and another for USB midi and attached call-backs to both (see the examples). Then send messages from your instrument to both Serial and USB.

So what I get is my sparkFun pro micro (5v 16MHz) frozen and unresponsive at boot.
Is there anything I am missing?

EDIT (code added):
```
#include <Wire.h>
#include <Adafruit_SSD1306.h>
#include <Adafruit_GFX.h>

#include <USB-MIDI.h>
USING_NAMESPACE_MIDI;

USBMIDI_CREATE_INSTANCE(1, MIDIusb);
// MIDI_CREATE_INSTANCE(HardwareSerial, Serial1, MIDIhw);

#define SDA 2
#define SCL 3
#define SCREEN_WIDTH 128 
#define SCREEN_HEIGHT 64 
#define OLED_RESET -1
#define SCREEN_ADDRESS 0x3C
#define SSD1306_NO_SPLASH
Adafruit_SSD1306 display(SCREEN_WIDTH, SCREEN_HEIGHT, &Wire, OLED_RESET);

#define FAD_A 21
#define FAD_B 20
uint_fast8_t fad_A_ch = 0;
uint_fast8_t fad_A_cc = 0;
bool fad_A_port = true;
uint_fast64_t fad_A_last = 0;
uint_fast8_t fad_B_ch = 0;
uint_fast8_t fad_B_cc = 1;
bool fad_B_port = false;
uint_fast64_t fad_B_last = 0;


void setup() {
  Serial.begin(115200);
  delay(1000);

  pinMode(FAD_A, INPUT);
  pinMode(FAD_B, INPUT);

  btn.attach(BTN_PIN, INPUT_PULLUP);
  btn.interval(10);

  MIDIusb.begin();
  MIDIusb.turnThruOff();
  // MIDIhw.begin();
  // MIDIhw.turnThruOff();

  while(!display.begin(SSD1306_SWITCHCAPVCC, SCREEN_ADDRESS)) {}
}

void loop() {
  MIDIusb.read();
  // MIDIhw.read();

  handleFaders();
  handleOLED();
  
}

void handleOLED(){
  display.clearDisplay();

  for (uint8_t i = 0; i<2; i++){
    display.setCursor(i*24, 32);
    display.setTextColor(SSD1306_WHITE);

    char buffer[4];
    if( i==0 ){
      sprintf(buffer, " %03d", fad_A_last);
    } else if ( i==1){
      sprintf(buffer, " %03d", fad_B_last);
    } 
    display.println(buffer);
  }
  display.display();
}

void handleFaders(){
  uint_fast64_t val_A = analogRead(FAD_A) / 8;
  uint_fast64_t val_B = analogRead(FAD_B) / 8;

  if (val_A != fad_A_last){
    controlChange(fad_A_ch, fad_A_cc, val_A, fad_A_port);
    fad_A_last = val_A;
  }
  if (val_B != fad_B_last){
    controlChange(fad_B_ch, fad_B_cc, val_B, fad_B_port);
    fad_B_last = val_B;
  }
}

void controlChange(uint8_t chan, uint8_t ctl, uint8_t val, bool port) {
  if (port){
    MIDIusb.sendControlChange(ctl, val, chan+1);
  } else {
    // MIDIhw.sendControlChange(ctl, val, chan+1);
}

```

**Reply by serlecu** [reply]
Nevermind, I found there is some kind of incompatibility with my implementation of an OLED screen.
